lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -507,48 +542,47 @@ def _post(self, json_data: JSON, _) -> Response:
                 status = "ok"
                 if "error" in u:
                     e = u["error"]
-                    status = e["type"]
-                    self.logger.debug(
-                        "{} ({}: {}) for id {} in index {}",
-                        u["status"],
-                        status,
-                        e["reason"],
-                        u["_id"],
-                        u["_index"],
-                    )
+                    status = e["reason"]
                     error_count += 1
-                cnt = report.get(status, 0)
-                report[status] = cnt + 1
+                report[status][u["_index"]] += 1
         except Exception as e:
             self.logger.exception(
-                "{}: exception {} occurred during the Elasticsearch request",
+                "{}: exception {} occurred during the Elasticsearch request: report {}",
                 klasname,
                 type(e).__name__,
+                report,
             )
             abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR")
 
-        self.logger.info(
-            "{}:dataset {}: {} successful document updates and {} failures",
-            klasname,
-            dataset,
-            count - error_count,
-            error_count,
-        )
+        summary = {"ok": count - error_count, "failure": error_count}
 
         # Let the subclass complete the operation
-        self.complete(dataset, json_data, error_count)
+        self.complete(dataset, json_data, summary)
 
-        # Return the report document as the success response, or abort with an
+        # Return the summary document as the success response, or abort with an
         # internal error if we weren't 100% successful. Some elasticsearch
         # documents may have been affected, but the client will be able to try
-        # again. TODO: switching to `pyesbulk` will automatically handle
-        # retrying on non-terminal errors, but this requires some cleanup
-        # work on the pyesbulk side.
+        # again.
+        #
+        # TODO: switching to `pyesbulk` will automatically handle retrying on
+        # non-terminal errors, but this requires some cleanup work on the
+        # pyesbulk side.
         if error_count > 0:
+            self.logger.error(
+                "{}:dataset {}: {} successful document updates and {} failures: {}",
+                klasname,
+                dataset,
+                count - error_count,
+                error_count,
+                report,
+            )
             abort(
                 HTTPStatus.INTERNAL_SERVER_ERROR,
                 message=f"{error_count:d} of {count:d} Elasticsearch document UPDATE operations failed",
-                data=report,
+                data=summary,
             )
 
-        return report
+        self.logger.info(
+            "{}:dataset {}: {} successful document updates", klasname, dataset, count
+        )
+        return summary
```

Comment:
It's pretty flexible in what it accepts. You can return a full Response object, or various formats the define one implicitly. Our existing code is a mixed bag of formats; a lot are using an explicit `jsonify(...)`, which is fine, but returning a dict will be implicitly jsonified, and the default status is 200. You can return a tuple with (status, data), too.

I didn't change the format of the return here, so I'm not going to worry about this; but if we want to think about forcing some sort of consistency in the future, arguably we should go for a full `Response` object on each API since that's the most flexible. (And Flask transformers will end up building one anyway.)